[docs] Add "extend component sizes and variants" section (#34050)
mui/material-ui: docs/data/joy/customization/themed-components/themed-components.md

}
```

## Default props

> The values specified in the theme `defaultProps` affect to all instances of the component:

Reviewer: ```suggestion
The values specified in the theme as `defaultProps` affect all instances of the component:
```

});
```

### Extend sizes

> Apply the styles when `ownerState.size` matches the new size. It is recommended to follow the existing T-shirt scale for consistency.

Reviewer: ```suggestion
Here's how you'd go for extending a component's available sizes.
We recommend following the in-use t-shirt sizing convention to maintain consistency with all the other props.
```

> For example, adding `xs` and `xl` sizes to the button would be:

Reviewer: ```suggestion

```

```js
extendTheme({
  components: {
    JoyButton: {
      styleOverrides: {
        root: ({ ownerState, theme }) => ({
          ...(ownerState.size === 'xs' && {
            // smallest size
            '--Icon-fontSize': '1rem',
            '--Button-gap': '0.25rem',
            minHeight: 'var(--Button-minHeight, 1.75rem)',
            fontSize: theme.vars.fontSize.sm,
            paddingBlock: '2px',
            paddingInline: '0.5rem',
          }),
          ...(ownerState.size === 'xl' && {
            // largest size
            '--Icon-fontSize': '2rem',
            '--Button-gap': '1rem',
            minHeight: 'var(--Button-minHeight, 4rem)',
            fontSize: theme.vars.fontSize.lg,
            paddingBlock: '0.5rem',
            paddingInline: '2rem',
          }),
        }),
      },
    },
  },
});
```

> :::info
> **TypeScript**
> 
> You need module augmentation to include the values to the `size` prop of the component.
> 
> The interface is in `{ComponentName}PropsSizeOverrides` format which exists in all Joy UI components:

Reviewer: ```suggestion
:::info
ℹ️ **TypeScript tip**:  You need module augmentation to include the values to the `size` prop of the component.

The interface is in `{ComponentName}PropsSizeOverrides` format which exists in all Joy UI components:
```

```tsx
// This part could be declare in your theme file
declare module '@mui/joy/Button' {
  interface ButtonPropsSizeOverrides {
    xs: true;
    xl: true;
  }
}

// typed-safe
<Button size="xs" />
<Button size="xl" />
```

:::

### Extend variants

> When you add new variant to a component, you are opting-out of the global variant feature which lets you take full control of the CSS properties related to color (`color`, `background` and `border`).

Reviewer: ```suggestion
By creating new component variants, you're automatically opting out of the [global variant feature](/joy-ui/main-features/global-variants/), which allows you to have fine-grain control of the color-related CSS properties (`color`, `background` and `border`).
```

```js
extendTheme({
  components: {
    JoySheet: {
      styleOverrides: {
        root: ({ ownerState, theme }) => ({
          ...(ownerState.variant === 'glass' && {
            color: theme.vars.palette.text.primary,
            background: 'rgba(255, 255, 255, 0.14)',
            backdropFilter: 'blur(5px)',
            border: '1px solid rgba(255, 255, 255, 0.3)',
            boxShadow: '0 4px 30px rgba(0, 0, 0, 0.1)',
          }),
        }),
      },
    },
  },
});
```

:::info

> **TypeScript**
> 
> You need module augmentation to include the values to the `variant` prop of the component.

Reviewer: ```suggestion
ℹ️ **TypeScript tip**:  You need module augmentation to include the values to the `size` prop of the component.
```

Author: Actually, this is needed rather than a tip. Should I move it out of the `:::info`?

Reviewer: Ah, that makes sense. Yeah, I think so!

We recommend following the in-use t-shirt sizing convention to maintain consistency with all the other props.

Here is an example of extending the button sizes to include `xs` and `xl` values:

```js
extendTheme({
  components: {
    JoyButton: {
      styleOverrides: {
        root: ({ ownerState, theme }) => ({
          ...(ownerState.size === 'xs' && {
            '--Icon-fontSize': '1rem',
            '--Button-gap': '0.25rem',
            minHeight: 'var(--Button-minHeight, 1.75rem)',
            fontSize: theme.vars.fontSize.xs,
            paddingBlock: '2px',
            paddingInline: '0.5rem',
          }),
          ...(ownerState.size === 'xl' && {
            '--Icon-fontSize': '2rem',
            '--Button-gap': '1rem',
            minHeight: 'var(--Button-minHeight, 4rem)',
            fontSize: theme.vars.fontSize.xl,
            paddingBlock: '0.5rem',
            paddingInline: '2rem',
          }),
        }),
      },
    },
  },
});
```

Then, you will be able to use those values on the button instances:

```jsx
<Button size="xs">Extra small</Button>
<Button size="xl">Extra large</Button>
```

:::info

> **Tip**: the properties for extending sizes should relate to density or dimension of the component. If you want to control the color-related properties, you should [extend the variants](#extend-variants) instead.

Reviewer 2: ```suggestion
The properties used for extending sizes should only relate to the density or the dimensions of the component. 
To learn how to extend color properties, check out the [Extend variants](#extend-variants) section in this document.
```

:::

#### TypeScript

You need module augmentation to include the values to the `size` prop of the component.

The interface is in `{ComponentName}PropsSizeOverrides` format which exists in all Joy UI components:

```tsx

> // This part could be declare in your theme file

Reviewer 2: ```suggestion
// This part could be declared in your theme file
```

declare module '@mui/joy/Button' {
  interface ButtonPropsSizeOverrides {
    xs: true;
    xl: true;
  }
}

// typed-safe
<Button size="xs" />
<Button size="xl" />
```

### Extend variants

> By creating new component variants, you're automatically opting out of the [global variant feature](/joy-ui/main-features/global-variants/), which allows you to have fine-grain control of the color-related CSS properties (`color`, `background` and `border`).

Reviewer 2: ```suggestion
The following code snippet shows how to extend component variants for color properties.
Note that by creating new variants, you're automatically opting out of the [global variant feature](/joy-ui/main-features/global-variants/), which gives you fine-grained control over CSS properties like `color`, `background`, and `border`.
```

> Here is an example of extending the sheet variant to include `glass` value:

Reviewer 2: ```suggestion
This example extends the Sheet variant to include a custom value named `glass`:
```

```js
extendTheme({
  components: {
    JoySheet: {
      styleOverrides: {
        root: ({ ownerState, theme }) => ({
          ...(ownerState.variant === 'glass' && {
            color: theme.vars.palette.text.primary,
            background: 'rgba(255, 255, 255, 0.14)',
            backdropFilter: 'blur(5px)',
            border: '1px solid rgba(255, 255, 255, 0.3)',
            boxShadow: '0 4px 30px rgba(0, 0, 0, 0.1)',
          }),
        }),
      },
    },
  },
});
```

> Then, you will be able to use the value on the sheet instances:

Reviewer 2: ```suggestion
Once the value is defined as above, you can make use of it directly on instances of the Sheet component:
```

```jsx
<Sheet variant="glass">Glassmorphism</Sheet>
```

#### TypeScript

> You need module augmentation to include the values to the `variant` prop of the component.

Reviewer 2: ```suggestion
Module augmentation is required to pass the values to the `variant` prop of the component.
```

> The interface is in `{ComponentName}PropsVariantOverrides` format which exists in all Joy UI components:

Reviewer 2: ```suggestion
The interface format is `{ComponentName}PropsSizeOverrides`, which is the same for all Joy UI components:
```

```tsx

> // This part could be declare in your theme file

Reviewer 2: ```suggestion
// This part could be declared in your theme file
```
